build(docs-infra): fix `autoLinkCode` to ignore docs without a path

sonukapoor opened the pull request:
Closes #36260

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [X] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #36260


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

sonukapoor committed:
build(docs-infra): fix `autoLinkCode` to ignore docs without a path

Previously, the auto linker generated links without an `href` when the
API was private. This commit fixes this by making sure that the `path`
of the document is not empty.

Closes #36260

petebacondarwin reviewed (changes requested):
While I think this is effective, it doesn't get around the fact that we are mentioning API items in the docs, which are not actually part of public API!
At the least I think that this PR ought to print a warning in such a case, rather than quietly ignoring it.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -77,7 +77,8 @@ module.exports = function autoLinkCode(getDocFromAlias) {
     return docs.length === 1 && !docs[0].internal &&
-        autoLinkCodeImpl.docTypes.indexOf(docs[0].docType) !== -1;
+        autoLinkCodeImpl.docTypes.indexOf(docs[0].docType) !== -1 &&
+        docs[0].path !== '';
```
Can you also add  comment explaining why a doc would not have a `path`?

petebacondarwin commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -75,10 +75,25 @@ module.exports = function autoLinkCode(getDocFromAlias) {
+   * that the docs are not `internal` and that the `docType` is supported. The `path`
+   * can be empty when the `API` is not public.
+   *
+   * @param {Array<Object>} docs An array of objects containing the doc details
```
Thanks for adding the JSDOC block. If you are going to add one param for `docs` then I think you should also document the `keyword` param too.

sonukapoor replied:
My bad. I added the second param after the JSDoc.

petebacondarwin commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -75,10 +75,25 @@ module.exports = function autoLinkCode(getDocFromAlias) {
+
+    var doc = docs[0];
+    if (doc.path === '') {
+      log.warn(`Doc path is empty for "${doc.id}" - link will not be generated for "${keyword}".`);
```
Can you mention in this warning that this is coming from the `autoLinkCode` post-processor, since there would be no other information given in the warning and it would be difficult for people to resolve this warning.

petebacondarwin reviewed (changes requested):
Great! Thanks for adding the logging. A couple of NITs.
Also, could you run the doc-gen locally to see how many warning we get with the current docs?

sonukapoor commented:
> Great! Thanks for adding the logging. A couple of NITs.
> Also, could you run the doc-gen locally to see how many warning we get with the current docs?

I see a lot - let me know how I can help to fix those:

```
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.exported" - link will not be generated for "exported".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.exported" - link will not be generated for "exported".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.exported" - link will not be generated for "exported".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.compilation" - link will not be generated for "compilation".
warn:    autoLinkCode: Doc path is empty for "core/ɵDepFlags.Value" - link will not be generated for "Value".
warn:    autoLinkCode: Doc path is empty for "core/ɵDepFlags.Value" - link will not be generated for "Value".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵRenderFlags.Update" - link will not be generated for "Update".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.exported" - link will not be generated for "exported".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "platform-browser/ɵBrowserDomAdapter.createElement()" - link will not be generated for "createElement()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/ɵBrowserDomAdapter.createElement()" - link will not be generated for "createElement()".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.nodes" - link will not be generated for "nodes".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.nodes" - link will not be generated for "nodes".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.selectors" - link will not be generated for "selectors".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵAttributeMarker.Template" - link will not be generated for "Template".
warn:    autoLinkCode: Doc path is empty for "core/ɵPlayState.Finished" - link will not be generated for "Finished".
warn:    autoLinkCode: Doc path is empty for "core/ɵPlayState.Finished" - link will not be generated for "Finished".
warn:    autoLinkCode: Doc path is empty for "core/ɵPlayState.Finished" - link will not be generated for "Finished".
warn:    autoLinkCode: Doc path is empty for "core/ɵPlayState.Finished" - link will not be generated for "Finished".
warn:    autoLinkCode: Doc path is empty for "core/ɵCurrencyIndex.Symbol" - link will not be generated for "Symbol".
warn:    autoLinkCode: Doc path is empty for "core/ɵCurrencyIndex.Symbol" - link will not be generated for "Symbol".
warn:    autoLinkCode: Doc path is empty for "core/ɵRenderFlags.Update" - link will not be generated for "Update".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵRenderFlags.Update" - link will not be generated for "Update".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵRenderFlags.Create" - link will not be generated for "Create".
warn:    autoLinkCode: Doc path is empty for "core/ɵAttributeMarker.Template" - link will not be generated for "Template".
warn:    autoLinkCode: Doc path is empty for "core/ɵLContext.native" - link will not be generated for "native".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵAttributeMarker.Template" - link will not be generated for "Template".
warn:    autoLinkCode: Doc path is empty for "core/ɵArgumentType.Dynamic" - link will not be generated for "Dynamic".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "platform-browser/ɵBrowserDomAdapter.dispatchEvent()" - link will not be generated for "dispatchEvent()".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵLContext.native" - link will not be generated for "native".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onChanges" - link will not be generated for "onChanges".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵAttributeMarker.Styles" - link will not be generated for "Styles".
warn:    autoLinkCode: Doc path is empty for "core/ɵAttributeMarker.Styles" - link will not be generated for "Styles".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵBrowserAnimationBuilder.build()" - link will not be generated for "build()".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.onInit" - link will not be generated for "onInit".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.nodes" - link will not be generated for "nodes".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵCodegenComponentFactoryResolver.resolveComponentFactory()" - link will not be generated for "resolveComponentFactory()".
warn:    autoLinkCode: Doc path is empty for "platform-browser/ɵDomSanitizerImpl" - link will not be generated for "ɵDomSanitizerImpl".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "platform-browser/animations/ɵAnimationRenderer.setProperty()" - link will not be generated for "setProperty()".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵCodegenComponentFactoryResolver.resolveComponentFactory()" - link will not be generated for "resolveComponentFactory()".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckOnce" - link will not be generated for "CheckOnce".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "core/ɵChangeDetectorStatus.CheckAlways" - link will not be generated for "CheckAlways".
warn:    autoLinkCode: Doc path is empty for "animations/browser/ɵWebAnimationsPlayer.time" - link will not be generated for "time".
warn:    autoLinkCode: Doc path is empty for "core/ɵRenderFlags.Update" - link will not be generated for "Update".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵViewDefinition.flags" - link will not be generated for "flags".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.First" - link will not be generated for "First".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵDirectiveDef.selectors" - link will not be generated for "selectors".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.compilation" - link will not be generated for "compilation".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.compilation" - link will not be generated for "compilation".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵComponentDef._" - link will not be generated for "_".
warn:    autoLinkCode: Doc path is empty for "core/ɵQueryBindingType.All" - link will not be generated for "All".
warn:    autoLinkCode: Doc path is empty for "core/ɵNgModuleTransitiveScopes.compilation" - link will not be generated for "compilation".
```

petebacondarwin commented:
I think we should look at where these have been used in the docs and probably remove the code if possible / or maybe block autolinking for those cases. Alternatively, if it appears that the docs are correct, we might want to add some of these items to the public API.

sonukapoor commented:
Sounds good. I will do that.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -75,9 +75,27 @@ module.exports = function autoLinkCode(getDocFromAlias) {
+      return false;
+    }
+
+    return !doc.internal && autoLinkCodeImpl.docTypes.indexOf(doc.docType) !== -1;
```
You can also do:

```suggestion
    return !doc.internal && autoLinkCodeImpl.docTypes.includes(doc.docType);
```

sonukapoor replied:
Done.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js:
```diff
@@ -75,9 +75,27 @@ module.exports = function autoLinkCode(getDocFromAlias) {
+    }
+
+    var doc = docs[0];
+    if (doc.path === '') {
```
@petebacondarwin, would it be worth marking these docs as private (i.e. have something like `doc.private === true`), so that this is more clear?

petebacondarwin replied:
It is complicated by the fact that the linked code can be a member of an enum API doc, and that the member is public while the containing doc is private.

It is likely that these docs are indeed already marked as private anyway. But even if they were no private then not having a `path` property precludes them from being linked to. So I think that this is the correct check.

gkalpak reviewed (approved):
:+1:

+1 for eliminating most of the warnings (one way or another).

sonukapoor commented:
@petebacondarwin / @gkalpak What exactly should be done about the warnings? I double-checked a few of them and they are indeed in private exported files such as `*_private_exports.ts`. Should I make an excel document for each warning and list where the file is exported from and then we will tackle each entry one at a time? We could probably then discuss if the item should be public or not and take action. Also, is this change going to be in this PR or in a new PR?

gkalpak commented:
I am fine fixing the current warnings in this PR or a follow-up, but I believe we should not have a wall of warnings, because they make real warnings hard to see. If they are going to be fixed in a different PR, then I'd vote for disabling the warnings in this PR and re-enabling them when the bulk of them are fixed on that follow-up PR.

sonukapoor commented:
@gkalpak / @petebacondarwin The sheet is ready for you both to review. I see already one item that can be fixed, e.g: we use `_` in some code guides to explain something related to code standards. The `autoLinkCode` tries to match that, to generate a link. We should add `_` to an `ignored keywords` array - I don't think that exists yet, I can create one, but let's first discuss what should be done with the remainder of the entries.  👍

gkalpak commented:
Thx, @sonukapoor, for putting [the spreadsheet][1] together.

From a quick look at it, I think we should put the following generic terms in an ignored list:
`_`, `All`, `build`, `compilation`, `Create`, `Dynamic`, `exported`, `Finished`, `First`, `flags`, `onChanges`, `onInit`, `native`, `nodes`, `selectors`, `Styles`, `Symbol`, `Template`, `time`, `Update`, `Value`

After that, we should investigate what the appropriate solution is for the following:
`CheckAlways`, `CheckOnce`, `createElement`, `dispatchEvent`, `resolveComponentFactory, `setProperty`

Finally, I could not find `ɵDomSanitizerImpl`. Where exactly is it mentioned in the docs?

[1]: https://docs.google.com/spreadsheets/d/1oKWcunWiU453RckgCx8ft6Tlr2YkUzhhWGv7_c7tf0E

gkalpak commented:
BTW, there is already an `ignoreGenericWords` custom filer [here][1].
<sub>Looking at that, I wonder why `Create` and `Value` are not filtered out 😕</sub>

[1]: https://github.com/angular/angular/blob/549ab2ecc3105810507442aba15577eedf31d7f3/aio/tools/transforms/angular-base-package/services/auto-link-filters/ignoreGenericWords.js

sonukapoor commented:
> BTW, there is already an `ignoreGenericWords` custom filer [here](https://github.com/angular/angular/blob/549ab2ecc3105810507442aba15577eedf31d7f3/aio/tools/transforms/angular-base-package/services/auto-link-filters/ignoreGenericWords.js).
> Looking at that, I wonder why `Create` and `Value` are not filtered out 😕

Maybe we are doing a case-sensitive comparison?

gkalpak commented:
No, [were are not][1].

[1]: https://github.com/angular/angular/blob/549ab2ecc3105810507442aba15577eedf31d7f3/aio/tools/transforms/angular-base-package/services/auto-link-filters/ignoreGenericWords.js#L9

petebacondarwin commented:
Also from looking at the sheet, we could get quite fair in ensuring that `member` docs are only linked if the linking text starts with a `.`.  Not quire sure the best way to achieve this though.

petebacondarwin commented:
And if we wanted to be super sophisticated we could consider the language of the code snippet, perhaps?

petebacondarwin commented:
I feel like the auto-linker should generally be quite conservative (risk-averse) and rather not link things that it should rather than link things that it shouldn't... WDYT?

sonukapoor commented:
@petebacondarwin / @gkalpak Are we OK with the file for me to continue to apply the fixes?

petebacondarwin commented:
Sorry not to get back to you. I think we could land this PR without the warning message for now. And then work on fixing the warnings. I would rather try to find some systematic ways to improve the matching rather than just ignoring all the keywords they trigger the warning. (`_` being a notable exception to this - we should ignore that).

petebacondarwin commented:
For example, how about changing this line: https://github.com/angular/angular/blob/master/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js#L61

to something like:

```ts
return foundValidDoc(filteredDocs) &&
              (filteredDocs[0].docType !== 'member' || word.includes('.')) ?
```

sonukapoor commented:
> Sorry not to get back to you. I think we could land this PR without the warning message for now. And then work on fixing the warnings. I would rather try to find some systematic ways to improve the matching rather than just ignoring all the keywords they trigger the warning. (`_` being a notable exception to this - we should ignore that).

I agree, should I remove the logging part from the PR?

sonukapoor commented:
> For example, how about changing this line:
> 
> https://github.com/angular/angular/blob/master/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js#L61
> 
> to something like:
> 
> ```ts
> return foundValidDoc(filteredDocs) &&
>               (filteredDocs[0].docType !== 'member' || word.includes('.')) ?
> ```

We would have to change this line too, right?

https://github.com/angular/angular/blob/master/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js#L49

petebacondarwin commented:
Yes I think so.

sonukapoor commented:
@petebacondarwin PR updated with the new condition.

petebacondarwin commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js:
```diff
@@ -126,6 +126,15 @@ describe('autoLinkCode post-processor', () => {
     expect(doc.renderedContent).toEqual('<code>MyClass</code>');
   });
 
+  it('should ignore code items match an API doc but has not path set',
```
```suggestion
  it('should ignore code items that match an API doc but have no path set',
```

petebacondarwin reviewed (approved):
This looks great! The last warning message is because we are displaying the private class in the inheritance hierarchy for the `Sanitizer` and `ɵDomSanitizerImpl` is a sub-class. We should do a separate fix to ensure that private classes do not appear in such hierarchies.

sonukapoor commented:
> This looks great! The last warning message is because we are displaying the private class in the inheritance hierarchy for the `Sanitizer` and `ɵDomSanitizerImpl` is a sub-class. We should do a separate fix to ensure that private classes do not appear in such hierarchies.

I will take care of that.

gkalpak reviewed (approved):
Awesome work, @sonukapoor 🚀

Can you please add some tests for the new `doc.docType === 'member' && !keyword.includes('.')` check?

I would also split these changes into 3 commits:
- One adding the `doc.path === ''` check.
- One adding the `doc.docType === 'member' && !keyword.includes('.')` check.
- One adding `_` to the `ignoredWords` list of the `ignoreGenericWords` fitler.

(Or feel free to keep them in one commit, but update the commit message to be more generic and describe all three fixes.)

sonukapoor commented:
> Awesome work, @sonukapoor 🚀
> 
> Can you please add some tests for the new `doc.docType === 'member' && !keyword.includes('.')` check?
> 
> I would also split these changes into 3 commits:
> 
> * One adding the `doc.path === ''` check.
> * One adding the `doc.docType === 'member' && !keyword.includes('.')` check.
> * One adding `_` to the `ignoredWords` list of the `ignoreGenericWords` fitler.
> 
> (Or feel free to keep them in one commit, but update the commit message to be more generic and describe all three fixes.)

Makes sense. I will work on it.

sonukapoor commented:
> * One adding the `doc.path === ''` check.

You mean `doc.path !== ''`  ?

gkalpak commented:
I mean this check:

https://github.com/angular/angular/blob/1157f2cae3a158ba1761ffc7c6787f8af66307b7/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.js#L98

sonukapoor committed:
build(docs-infra): Ensures that only member docs are linked

This commit ensures that `member` docs are only linked if the linking
text contains `.`.

sonukapoor committed:
build(docs-infra): add `_` to ignored ignoreGenericWords

Previously, when a document included `_`, the autoLinker will try to
generate a link, e.g from `core/ɵComponentDef._`. This commit adds it
to the ignored words to prevent that.

sonukapoor commented:
@gkalpak Changes done. Please review.

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js:
```diff
@@ -135,6 +135,16 @@ describe('autoLinkCode post-processor', () => {
+       aliasMap.addDoc({docType: 'class', id: 'MyClass', aliases: ['MyClass'], path: 'a/b/myclass'});
+       aliasMap.addDoc(
+           {docType: 'member', id: 'MyEnum', aliases: ['MyEnum'], path: 'a/b/c'});
+       const doc = {docType: 'test-doc', renderedContent: '<code>MyClass</code>'};
```
AFAICT, this test would pass even without the changes in this commit. We need a test thaty would fail without the changes and passes with the changes (e.g. mentioning `MyEnum` in the code block).

sonukapoor replied:
Correct, but It's making sure that a link is not generated for `MyEnum`.

sonukapoor replied:
Ok. Will revise

gkalpak commented on aio/tools/transforms/angular-base-package/services/auto-link-filters/ignoreGenericWords.spec.js:
```diff
@@ -16,6 +16,12 @@ describe('ignoreGenericWords', () => {
     expect(ignoreGenericWords(docs, words, 2)).toEqual(docs);
   });
+
+  it('should ignore `_` in all docs', () => {
```
This test is not necessary, but I don't mind one extra test :grin:

sonukapoor replied:
Will remove 😄

sonukapoor committed:
fixup!: build(docs-infra): add _ to ignored ignoreGenericWords

sonukapoor committed:
fixup!: build(docs-infra): Ensures that only member docs are linked

gkalpak commented on aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js:
```diff
@@ -137,12 +137,20 @@ describe('autoLinkCode post-processor', () => {
        processor.$process([doc]);
-       expect(doc.renderedContent).toEqual('<code><a href="a/b/myclass" class="code-anchor">MyClass</a></code>');
+       expect(doc.renderedContent).toEqual('<code><a href="a/b/c" class="code-anchor">MyEnum.Value</a></code>');
      });
```
There is already a test that verifies that: https://github.com/angular/angular/blob/5b3b5cac32dcc83baa97b79b3660a7cd86fa8e6b/aio/tools/transforms/angular-base-package/post-processors/auto-link-code.spec.js#L41-L54

Maybe you could move it here instead of adding an (essentially) identical one.
(The `doc` in the existing test is also "more correct", as it has `id: 'MyEnum.Value'` and `aliases: ['Value', 'MyEnum.Value']`.)

sonukapoor replied:
True. In that case, I will just remove mine. Is that OK?

sonukapoor committed:
fixup!: build(docs-infra): Ensures that only member docs are linked

gkalpak reviewed (approved):
LGTM ✨

sonukapoor commented:
@gkalpak, Can you please remove the  `PR action: cleanup` and add `PR action: merge`?
